### PR TITLE
Increase flexibility on which offset to start for kafka adapter, when previous offset not found or out of range

### PR DESCRIPTION
**Pull request by zliu41**
Before: if previous offset not found, it always starts at the earliest offset; if previous offset is out of range, either start at earliest offset or skip partition.

Now:
- If previous offset not found, may start at earliest or latest, or skip. Controlled by "bootstrap.with.offset".
- If previous offset out of range, may start at earliest or latest, or skip, or start at the nearest offset (if it's closer to earliest, start at earliest, otherwise start at latest). Controlled by "reset.on.offset.out.of.range".

### CODE REVIEW

**Comment by zliu41** [comment]
@ibuenros please take a look. Thanks.

**Comment by zliu41** [comment]
Updated, since `switch(string)` not supported in Java 6. Changed to if-else.

File: gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java [hunk 1]
```diff
@@ -377,31 +383,43 @@ private WorkUnit getWorkUnitForTopicPartition(KafkaWrapper kafkaWrapper, KafkaPa
...
 
     } catch (PreviousOffsetNotFoundException e) {
-      LOG.warn(String
-          .format(
-              "Previous offset for topic %s, partition %s does not exist. This partition will start from the earliest offset: %d",
-              partition.getTopicName(), partition.getId(), offsets.getEarliestOffset()));
-      offsets.startAtEarliestOffset();
+      String offsetNotFoundMsg = String.format("Previous offset for partition %s does not exist. ", partition);
+      String offsetOption = state.getProp(BOOTSTRAP_WITH_OFFSET, DEFAULT_BOOTSTRAP_WITH_OFFSET).toLowerCase();
+      if (offsetOption.equals(LATEST_OFFSET)) {
```
Comment (ibuenros) [inline gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java]: should this be equalsIgnoreCase? It makes it more user friendly.

**Comment by ibuenros** [comment]
Just one comment.

File: gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java (hunk 1 above)
Comment (zliu41) [inline gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java]: Previous line has `.toLowerCase()`.

**Comment by ibuenros** [comment]
LGTM

**Pull request merged**